### PR TITLE
LF: explicitly mark nodes if they correspond to a "by key" operation.

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
Currently we track the list of node that correspond to a "by key"
operation in the Transaction metadata. We move this information into
the nodes themself.

This is a preparatory work to include this information in the serialization 
format for  transactions

This advances the state of  #7622

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala [hunk 1]
```diff
@@ -281,7 +281,7 @@ private[kvutils] class TransactionCommitter(
       .fold((true, keys)) {
         case (
             (allUnique, existingKeys),
-            (_, exe @ Node.NodeExercises(_, _, _, _, _, _, _, _, _, _, _, _, _)))
+            (_, exe @ Node.NodeExercises(_, _, _, _, _, _, _, _, _, _, _, _, _, _)))
```
Comment (hurryabit) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:284]: This looks horrible. Can't we do something like this?
```suggestion
            (_, exe: Node.NodeExercises))
```
We have the same in many other places too. We should clean this up a bit in a separate PR. I'm happy to do it myself.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala [hunk 2]
```diff
@@ -354,7 +354,8 @@ object TransactionCoder {
           else if (txVersion precedes minContractKeyInFetch)
             Left(DecodeError(s"$txVersion is too old to support NodeFetch's `key` field"))
           else decodeKeyWithMaintainers(decodeCid, protoFetch.getKeyWithMaintainers).map(Some(_))
-        } yield (ni, NodeFetch(c, templateId, None, actingParties, signatories, stakeholders, key))
+        } yield
+          (ni, NodeFetch(c, templateId, None, actingParties, signatories, stakeholders, key, false))
```
Comment (hurryabit) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala:358]: Why do we always set this to `false` during decoding? Does the encoded version not contain the required information? Same question a bit further down.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala [hunk 3]
```diff
@@ -501,11 +501,11 @@ object GenTransaction extends value.CidContainer3[GenTransaction] {
           node match {
             case Node.NodeCreate(_, c, _, _, _, Some(key)) =>
               state.created(globalKey(c.template, key.key.value))
-            case Node.NodeExercises(_, tmplId, _, _, true, _, _, _, _, _, _, _, Some(key)) =>
+            case Node.NodeExercises(_, tmplId, _, _, true, _, _, _, _, _, _, _, Some(key), _) =>
               state.consumed(globalKey(tmplId, key.key.value))
-            case Node.NodeExercises(_, tmplId, _, _, false, _, _, _, _, _, _, _, Some(key)) =>
+            case Node.NodeExercises(_, tmplId, _, _, false, _, _, _, _, _, _, _, Some(key), _) =>
```
Comment (hurryabit) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala:506]: We need to improve this. A constructor pattern with so many positional arguments makes me cringe. Let's do it in a separate PR though.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala [hunk 4]
```diff
@@ -240,6 +247,7 @@ object Node {
       children: ImmArray[Nid],
       exerciseResult: Option[Val],
       key: Option[KeyWithMaintainers[Val]],
+      override val byKey: Boolean,
```
Comment (hurryabit) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:250]: I assume there's some invariant here we should document. I would guess that when `key` is `None` then `byKey` must be false. Does the inverse also hold? If not, why do we do store key information in the node if we didn't exercise by key?

Either way, if the presence of `key` means we've exercise by key, we don't need `byKey` at all. If `key` can be present even when we've exercised by contract id, we might want to put the boolean for `byKey` into `key`.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala [hunk 5]
```diff
@@ -196,7 +202,7 @@ object Node {
       with NodeInfo.Create {
 
     override def templateId: TypeConName = coinst.template
-
+    override def byKey: Boolean = false
```
Comment (hurryabit) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:205]: I find this confusing. Why would somebody ask a create note if it was "by key"?

**Review by hurryabit — APPROVED** [review]
I totally agree with the intent of the PR and the general implementation. I have questions about some details though. Many thanks.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala:358]: The encoded version does not contain the required information.
We have to add the flag to `transaction.proto`. I will do that in a following up PR.  
This PR is concerned with replacing the field `PartialTransaction#byKeyNodes` by the `byKey` flag in `GenNode`.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:250]: The invariant is `!byKey || key.isDefined`.

The reason why the `byKey` is not inside the `key` option is that this is a more drastic change, i.e. this will at least double the size of the PR. It is also a change of API that will impact ledger implementers. 

I you wish I can do it in another PR.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala (hunk 5 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:205]: There is an example in the PR, in `EngineTest.scala`
```
 private[this] def byKeyNodes[Nid, Cid](tx: VersionedTransaction[Nid, Cid]) =
    tx.nodes.collect { case (nodeId, node) if node.byKey => nodeId }.toSet
```

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:284]: In this case you have to write something like: 

```(_, exe: Node.NodeExercises[_, ContractId, VersionnedValue[ContractId]))```
or using type alias 
```(_, exe: Node.NodeExercises.WithTxValue[_, ContractId]))```

I really dislike matching with parametric types.  
But if you whish I will do the "cleanup".

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala (hunk 5 above)
Comment (hurryabit) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:205]: Oh, so it is mostly because of this old structure where the metadata contained information on whether a node was "by key" or not. That's not pretty but acceptable.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala (hunk 4 above)
Comment (hurryabit) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:250]: I think I do now understand why we need the `byKey` field where it is. Let's keep it there and add a comment documenting the invariant between `key` and `byKey`.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala (hunk 2 above)
Comment (hurryabit) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala:358]: Thanks for the explanation.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala (hunk 1 above)
Comment (hurryabit) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:284]: Can we do the following?
```scala
(_, exe: Node.Exercises[_, _, _])
```
I mean, we don't really care about the actual type arguments, do we?

Comment (remyhaemmerle-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:284]: I suspect in some place this broken the type inference. 
In this case, it seems to compile. However, this definitively broke the type inference of Intellij :-(

Comment (hurryabit) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:284]: That's weird that it breaks (IntelliJ's) type inference. The pattern `Node.NodeExercises(_, _, _, _, _, _, _, _, _, _, _, _, _, _)` doesn't make any statements about types whatsoever. Anyway, let's not touch this then for now.
